### ISSUE TITLE
getServerSideProps result is not passed to initial page request for client side app.

### ISSUE THREAD

**Issue by JorgeSivil** [zeit/next.js#11895]
# Bug report

First of all, I am pretty new to this JS world so forgive me and point me to learning resources if you feel I am not understanding something.

My problem is that the page is correctly rendered in server side (as it shows the Network inspector in Firefox) but then in client side the props are not initialized correctly for _app.js (which seeds a Context's state).

This works OK after changing pages with <Link>, so I get my username printed in the first page.

## Describe the bug

Even if I get the correct HTML, React app is executed and overwriting my server-side HTML.
So I get the HTML result, then React app is executed, since it doesn't have the getServerSideProps() result, my context lacks state, so React is drawing incorrect information.

## To Reproduce

_app.js

```
import React, {useContext} from "react";
import {UserContext} from "../lib/Context/UserContext"

export default function RankerApp({Component, pageProps}) {
    const [userState, setUserState] = useContext(UserContext)();
    const [listState, setListState] = useContext(ListContext)();

    if (pageProps.serverContext && pageProps.serverContext.user) {
        setUserState({...userState, ...pageProps.serverContext.user})
    }
    delete pageProps['serverContext'];

    return (
        <ListContext.Provider value={[listState, setListState]}>
            <UserContext.Provider value={[userState, setUserState]}>
                <Component {...pageProps} />
            </UserContext.Provider>
        </ListContext.Provider>
    );
}
```

lib/Context/UserContext

```
import {createContext, useState} from 'react';

const initialState = {
    loggedInUserId: undefined,
    users: {}
}

const userContextUpdater = () => {
    let [state, setState] = useState(initialState)
    return [state, setState];
}

export const UserContext = createContext(userContextUpdater);
```

pages/contextTest.js

```
import React, {useContext} from 'react';
import {UserContext} from "../lib/Context/UserContext";
import Link from "next/link";
//import UserContextSeeder from "../lib/User/UserContextSeeder";

export default function ContextPage1(props) {

    let [userState, setUserState] = useContext(UserContext);

    const currentLoggedInUserData = userState.loggedInUserId ? userState.users[userState.loggedInUserId] : null;
    const currentLoggedInUserName = currentLoggedInUserData ? currentLoggedInUserData.userName : 'Guest';

    return (
        <div>
            <h1>Page 1</h1>
            <Link href="contextTest2">
                <a>Page 2</a>
            </Link>
            <h1>Current Logged In User: {currentLoggedInUserName}</h1>
        </div>
    );
}

export async function getServerSideProps(context) {
    let userContext = {};
    // await UserContextSeeder(context).then(context => {userContext = context}); commented for simplicty
    userContext = {loggedInUserId: 222, users: {222: {userName: "Jorge"}}; // suppose this is the result
    return {
        props: {
            serverContext: {
                user: {...userContext}
            }
        },
    }
}
```

contextTest2.js

```
import React, {useContext} from 'react';
import {UserContext} from "../lib/Context/UserContext";
import Link from "next/link";

export default function ContextPage2(props) {

    let [userState, setUserState] = useContext(UserContext);

    const changeContext = () => {
        setUserState({...userState, value: userState.value + 1});
    }

    return (
        <div>
            <h1>Page 2</h1>
            <Link href="contextTest">
                <a>Page 1</a>
            </Link>
            <button onClick={changeContext}>changeContext()</button>
            <h1>{userState.value}</h1>
        </div>
    );
}
```

## Expected behavior

The page should have the correct username output.

## Screenshots

![image](https://user-images.githubusercontent.com/12187148/79277931-25e70700-7e81-11ea-933b-322e550dd3eb.png)

## System information
```

{
  "name": "ranker-v3",
  "version": "0.0.1",
  "main": "index.js",
  "license": "MIT",
  "scripts": {
    "dev": "next",
    "build": "next build",
    "start": "next start"
  },
  "dependencies": {
    "@zeit/next-sass": "^1.0.1",
    "next": "^9.3.4",
    "node-fetch": "^2.6.0",
    "node-sass": "^4.13.1",
    "react": "^16.13.0",
    "react-dom": "^16.13.0",
    "webpack": "^4.42.1"
  }
}
```

## Additional context

https://stackoverflow.com/questions/61217361/how-does-next-js-server-side-rendering-works-when-result-hits-client
https://github.com/zeit/next.js/discussions/11871

Resulting HTML in SSR
```

<!DOCTYPE html>
<html>
<head>
    <style data-next-hide-fouc="true">body {
        display: none
    }</style>
    <noscript data-next-hide-fouc="true">
        <style>body {
            display: block
        }</style>
    </noscript>
    <meta charSet="utf-8"/>
    <meta name="viewport" content="width=device-width"/>
    <meta name="next-head-count" content="2"/>
    <link rel="preload" href="/_next/static/development/pages/contextTest.js?ts=1586897061318" as="script"/>
    <link rel="preload" href="/_next/static/development/pages/_app.js?ts=1586897061318" as="script"/>
    <link rel="preload" href="/_next/static/runtime/webpack.js?ts=1586897061318" as="script"/>
    <link rel="preload" href="/_next/static/runtime/main.js?ts=1586897061318" as="script"/>
    <noscript id="__next_css__DO_NOT_USE__"></noscript>
</head>
<body>
<div id="__next">
    <div><h1>Page 1</h1><a href="contextTest2">Page 2</a>
        <h1>Current Logged In User: <!-- -->Jorge Sivil Ranker</h1></div>
</div>
<script src="/_next/static/development/dll/dll_fe850393d08f72bfe67b.js?ts=1586897061318"></script>
<script id="__NEXT_DATA__" type="application/json">{
    "props": {
        "pageProps": {},
        "__N_SSP": true
    },
    "page": "/contextTest",
    "query": {},
    "buildId": "development",
    "isFallback": false,
    "gssp": true
}</script>
<script nomodule="" src="/_next/static/runtime/polyfills.js?ts=1586897061318"></script>
<script async="" data-next-page="/contextTest"
        src="/_next/static/development/pages/contextTest.js?ts=1586897061318"></script>
<script async="" data-next-page="/_app" src="/_next/static/development/pages/_app.js?ts=1586897061318"></script>
<script src="/_next/static/runtime/webpack.js?ts=1586897061318" async=""></script>
<script src="/_next/static/runtime/main.js?ts=1586897061318" async=""></script>
<script src="/_next/static/development/_buildManifest.js?ts=1586897061318" async=""></script>
<script src="/_next/static/development/_ssgManifest.js?ts=1586897061318" async=""></script>
</body>
</html>

```

**Comment by JorgeSivil** [comment]
Ok so I dug more and found the way NextJs was hydrating the app, and then I found the fault was because of

`delete pageProps['serverContext'];`

So I changed to this:

```
if (pageProps.serverContext && pageProps.serverContext.user && Object.keys(userState.users).length === 0 && userState.users.constructor === Object) {
        setUserState({...userState, ...pageProps.serverContext.user})
    }
```

And I got the data client side! Had to make the extra check (going to do it cleaner) because it entered a loop otherwise, because of modifying the state. So I have to make sure this gets only set once.
